File: workflow/gee/land_cover/consolidate_safe_glc_fcs30d_exports.py
"""Validate and download a complete safe GLC-FCS30D watershed run."""

from __future__ import annotations

import argparse
import csv
import json
import math
import sys
from pathlib import Path
from typing import Any


try:
    import ee
except ImportError:
    ee = None


from run_safe_glc_fcs30d_exports import (
    DEFAULT_EXACT_MAX_WORK,
    DEFAULT_PROJECT,
    DEFAULT_RUN_ROOT,
    DEFAULT_SAMPLE_POINTS,
    GLC_CLASSES,
    METADATA_PROPERTIES,
    YEARS,
    child_asset_ids,
    load_sites,
    plan_tasks,
)


OUTPUT_COLUMNS = (
    *METADATA_PROPERTIES,
    "Year",
    "source",
    "LC_ID",
    "Area_m2",
    "extraction_method",
    "sample_count",
    "sample_n",
    "sample_fraction",
    "sample_standard_error",
    "requested_sample_n",
    "sampling_seed",
    "polygon_area_m2",
    "native_pixel_estimate",
    "effective_pixel_band_time",
    "glc_geometry_simplified_m",
    "glc_simplification_area_error_pct",
)


def parse_args() -> argparse.Namespace:
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument("--run-root", type=Path, default=DEFAULT_RUN_ROOT)
    parser.add_argument("--manifest", type=Path)
    parser.add_argument("--project", default=DEFAULT_PROJECT)
    parser.add_argument("--run-label", required=True)
    parser.add_argument("--output-folder")
    parser.add_argument("--method", choices=("auto", "exact", "sample"), default="auto")
    parser.add_argument("--sample-points", type=int, default=DEFAULT_SAMPLE_POINTS)
    parser.add_argument("--exact-max-work", type=float, default=DEFAULT_EXACT_MAX_WORK)
    parser.add_argument("--expected-site-count", type=int)
    parser.add_argument("--output", type=Path)
    parser.add_argument(
        "--download",
        action="store_true",
        help="Download only if every expected per-site asset is complete.",
    )
    args = parser.parse_args()
    args.manifest = args.manifest or args.run_root / "payload_manifest.csv"
    args.output_folder = args.output_folder or (
        f"projects/{args.project}/assets/glc_fcs30d_safe_{args.run_label}"
    )
    return args


def normalize_row(properties: dict[str, Any]) -> dict[str, Any]:
    row = dict(properties)
    row["site_id"] = str(row.get("site_id", ""))
    row["Year"] = int(row["Year"])
    row["LC_ID"] = int(row["LC_ID"])
    for name in (
        "Area_m2",
        "sample_count",
        "sample_n",
        "sample_fraction",
        "sample_standard_error",
        "requested_sample_n",
        "sampling_seed",
        "polygon_area_m2",
        "native_pixel_estimate",
        "effective_pixel_band_time",
    ):
        row[name] = float(row[name])
    return row


def validate_asset_rows(rows: list[dict[str, Any]], plan: Any) -> dict[str, Any]:
    expected_count = len(YEARS) * len(GLC_CLASSES)
    if len(rows) != expected_count:
        raise RuntimeError(
            f"{plan.asset_id} has {len(rows)} rows; expected {expected_count}."
        )
    keys = [(row["Year"], row["LC_ID"]) for row in rows]
    expected_keys = {(year, class_id) for year in YEARS for class_id in GLC_CLASSES}
    if len(keys) != len(set(keys)) or set(keys) != expected_keys:
        raise RuntimeError(f"Incomplete or duplicate year/class keys in {plan.asset_id}.")
    if {row["site_id"] for row in rows} != {plan.site.site_id}:
        raise RuntimeError(f"Wrong site_id in {plan.asset_id}.")
    expected_method = (
        "native_30m_exact"
        if plan.method == "exact"
        else "deterministic_point_sample"
    )
    if {row.get("extraction_method") for row in rows} != {expected_method}:
        raise RuntimeError(f"Wrong extraction_method in {plan.asset_id}.")

    minimum_sample_n: float | None = None
    maximum_fraction_error = 0.0
    for year in YEARS:
        year_rows = [row for row in rows if row["Year"] == year]
        area_sum = sum(row["Area_m2"] for row in year_rows)
        if not math.isfinite(area_sum) or area_sum <= 0:
            raise RuntimeError(f"Non-positive class area for {plan.site.site_id}, {year}.")
        if plan.method == "sample":
            sample_sizes = {row["sample_n"] for row in year_rows}
            if len(sample_sizes) != 1:
                raise RuntimeError(f"Inconsistent sample_n for {plan.site.site_id}, {year}.")
            sample_n = sample_sizes.pop()
            minimum_sample_n = (
                sample_n
                if minimum_sample_n is None
                else min(minimum_sample_n, sample_n)
            )
            if sample_n < 0.99 * plan.sample_points:
                raise RuntimeError(
                    f"Only {sample_n:,.0f}/{plan.sample_points:,} requested samples "
                    f"were classified for {plan.site.site_id}, {year}."
                )
            count_sum = sum(row["sample_count"] for row in year_rows)
            if not math.isclose(count_sum, sample_n, rel_tol=0, abs_tol=0.5):
                raise RuntimeError(
                    f"Class counts do not sum to sample_n for {plan.site.site_id}, {year}."
                )
            fraction_error = abs(
                sum(row["sample_fraction"] for row in year_rows) - 1
            )
            maximum_fraction_error = max(maximum_fraction_error, fraction_error)
            if fraction_error > 1e-6:
                raise RuntimeError(
                    f"Sample fractions do not sum to one for {plan.site.site_id}, {year}."
                )
            polygon_area = year_rows[0]["polygon_area_m2"]
            if not math.isclose(area_sum, polygon_area, rel_tol=1e-6, abs_tol=1):
                raise RuntimeError(
                    f"Sampled class areas do not sum to watershed area for "
                    f"{plan.site.site_id}, {year}."
                )
    return {
        "site_id": plan.site.site_id,
        "method": plan.method,
        "rows": len(rows),
        "minimum_sample_n": minimum_sample_n,
        "maximum_fraction_sum_error": maximum_fraction_error,
    }


def download_asset(plan: Any) -> tuple[list[dict[str, Any]], dict[str, Any]]:
    response = ee.FeatureCollection(plan.asset_id).getInfo()
    rows = [
        normalize_row(dict(feature.get("properties") or {}))
        for feature in response.get("features", [])
    ]
    return rows, validate_asset_rows(rows, plan)


def main() -> None:
    args = parse_args()
    if ee is None:
        raise RuntimeError(
            "The Earth Engine Python API is required. Install earthengine-api, "
            "authenticate, and rerun."
        )
    sites = load_sites(args.manifest)
    if args.expected_site_count is not None and len(sites) != args.expected_site_count:
        raise RuntimeError(f"Found {len(sites)} sites; expected {args.expected_site_count}.")
    plans = plan_tasks(
        sites=sites,
        method=args.method,
        sample_points=args.sample_points,
        exact_max_work=args.exact_max_work,
        run_label=args.run_label,
        output_folder=args.output_folder,
    )
    ee.Initialize(project=args.project)
    available = child_asset_ids(args.output_folder)
    missing = [plan.asset_id for plan in plans if plan.asset_id not in available]
    status = {
        "expected_assets": len(plans),
        "complete_assets": len(plans) - len(missing),
        "missing_assets": len(missing),
        "expected_sites": len(sites),
        "expected_rows": len(sites) * len(YEARS) * len(GLC_CLASSES),
    }
    print("Safe GLC consolidation status:", json.dumps(status, sort_keys=True))
    if not args.download:
        print("Read-only status complete; use --download only when all assets exist.")
        return
    if missing:
        raise RuntimeError(
            f"Refusing a partial download; {len(missing)} assets are missing.\n"
            + "\n".join(missing[:10])
        )

    all_rows: list[dict[str, Any]] = []
    asset_qa = []
    for index, plan in enumerate(plans, start=1):
        rows, qa = download_asset(plan)
        all_rows.extend(rows)
        asset_qa.append(qa)
        print(f"Downloaded and validated {index}/{len(plans)}: {plan.asset_id}")
    combined_keys = [
        (row["site_id"], row["Year"], row["LC_ID"]) for row in all_rows
    ]
    if len(combined_keys) != len(set(combined_keys)):
        raise RuntimeError("Duplicate site/year/class keys occur across assets.")
    if len(all_rows) != status["expected_rows"]:
        raise RuntimeError("The combined row count is incomplete.")

    output = args.output or (
        args.run_root / f"glc_fcs30d_safe_combined_{args.run_label}.csv"
    )
    output.parent.mkdir(parents=True, exist_ok=True)
    all_rows.sort(key=lambda row: (row["site_id"], row["Year"], row["LC_ID"]))
    with output.open("w", newline="", encoding="utf-8") as handle:
        writer = csv.DictWriter(handle, fieldnames=OUTPUT_COLUMNS, extrasaction="ignore")
        writer.writeheader()
        writer.writerows(all_rows)
    qa_path = output.with_suffix(".qa.json")
    qa = {
        **status,
        "combined_rows": len(all_rows),
        "output": str(output),
        "assets": asset_qa,
    }
    qa_path.write_text(json.dumps(qa, indent=2, sort_keys=True) + "\n", encoding="utf-8")
    print("Combined output:", output)
    print("QA receipt:", qa_path)


if __name__ == "__main__":
    try:
        main()
    except Exception as exc:
        print(f"SAFE GLC CONSOLIDATION ERROR: {exc}", file=sys.stderr)
        raise

File: workflow/gee/land_cover/consolidate_safe_glc_fcs30d_exports.R
### Setup

annual_glc_script_path <- function() {
  source_files <- vapply(sys.frames(), function(frame) {
    if (is.null(frame$ofile)) "" else as.character(frame$ofile)
  }, character(1))
  source_files <- source_files[nzchar(source_files)]
  if (length(source_files)) {
    return(normalizePath(tail(source_files, 1)))
  }
  file_arg <- grep("^--file=", commandArgs(FALSE), value = TRUE)
  if (length(file_arg)) {
    return(normalizePath(sub("^--file=", "", file_arg[[1]])))
  }
  normalizePath(
    "workflow/gee/land_cover/consolidate_safe_glc_fcs30d_exports.R"
  )
}

source(file.path(
  dirname(annual_glc_script_path()),
  "glc_consolidation_helpers.R"
))

glc_output_columns <- c(
  glc_metadata,
  "Year",
  "source",
  "LC_ID",
  "Area_m2",
  "extraction_method",
  "sample_count",
  "sample_n",
  "sample_fraction",
  "sample_standard_error",
  "requested_sample_n",
  "sampling_seed",
  "polygon_area_m2",
  "native_pixel_estimate",
  "effective_pixel_band_time",
  "glc_geometry_simplified_m",
  "glc_simplification_area_error_pct",
  glc_alias_columns
)
glc_sample_closure_tolerance <- 1e-6
glc_exact_boundary_tolerance <- 0.01

### Annual output checks

glc_normalize_row <- function(properties) {
  row <- properties
  row$site_id <- as.character(row$site_id %||% "")
  row$Year <- as.integer(row$Year)
  row$LC_ID <- as.integer(row$LC_ID)
  numeric_columns <- c(
    "Area_m2",
    "sample_count",
    "sample_n",
    "sample_fraction",
    "sample_standard_error",
    "requested_sample_n",
    "sampling_seed",
    "polygon_area_m2",
    "native_pixel_estimate",
    "effective_pixel_band_time"
  )
  for (name in numeric_columns) {
    row[[name]] <- as.numeric(row[[name]] %||% NA_real_)
  }
  row
}

glc_validate_asset_rows <- function(rows, plan) {
  expected_count <- length(glc_years) * length(glc_classes)
  if (length(rows) != expected_count) {
    stop(
      plan$asset_id,
      " has ",
      length(rows),
      " rows; expected ",
      expected_count
    )
  }
  keys <- vapply(
    rows,
    function(row) paste(row$Year, row$LC_ID, sep = ":"),
    character(1)
  )
  expected_keys <- as.vector(
    outer(glc_years, glc_classes, paste, sep = ":")
  )
  if (anyDuplicated(keys) || !setequal(keys, expected_keys)) {
    stop("Incomplete or duplicate year and class rows in ", plan$asset_id)
  }
  site_ids <- unique(vapply(rows, `[[`, character(1), "site_id"))
  if (!identical(site_ids, plan$site$site_id)) {
    stop("Wrong site ID in ", plan$asset_id)
  }
  expected_method <- if (plan$method == "exact") {
    "native_30m_exact"
  } else {
    "deterministic_local_equal_area_points_v1"
  }
  methods <- unique(vapply(
    rows,
    function(row) as.character(row$extraction_method %||% ""),
    character(1)
  ))
  if (!identical(methods, expected_method)) {
    stop("Wrong extraction method in ", plan$asset_id)
  }

  minimum_sample_n <- NULL
  maximum_fraction_error <- 0
  maximum_area_closure_error <- 0
  minimum_sample_fraction <- glc_check_sample_fraction(
    plan$minimum_sample_fraction %||% 0.99
  )
  for (year in glc_years) {
    year_rows <- Filter(function(row) row$Year == year, rows)
    area_sum <- sum(vapply(year_rows, `[[`, numeric(1), "Area_m2"))
    if (!is.finite(area_sum) || area_sum <= 0) {
      stop("Non-positive class area for ", plan$site$site_id, ", ", year)
    }
    polygon_areas <- unique(vapply(
      year_rows,
      `[[`,
      numeric(1),
      "polygon_area_m2"
    ))
    if (
      length(polygon_areas) != 1L ||
        !is.finite(polygon_areas[[1]]) ||
        polygon_areas[[1]] <= 0
    ) {
      stop("Invalid watershed area for ", plan$site$site_id, ", ", year)
    }
    polygon_area <- polygon_areas[[1]]
    area_closure_error <- abs(area_sum - polygon_area) / polygon_area
    maximum_area_closure_error <- max(
      maximum_area_closure_error,
      area_closure_error
    )
    if (plan$method == "sample") {
      sample_sizes <- unique(vapply(
        year_rows,
        `[[`,
        numeric(1),
        "sample_n"
      ))
      if (length(sample_sizes) != 1L) {
        stop("Inconsistent sample size for ", plan$site$site_id, ", ", year)
      }
      sample_n <- sample_sizes[[1]]
      minimum_sample_n <- if (is.null(minimum_sample_n)) {
        sample_n
      } else {
        min(minimum_sample_n, sample_n)
      }
      if (sample_n < minimum_sample_fraction * plan$sample_points) {
        stop("Too few classified samples for ", plan$site$site_id, ", ", year)
      }
      count_sum <- sum(vapply(
        year_rows,
        `[[`,
        numeric(1),
        "sample_count"
      ))
      if (abs(count_sum - sample_n) > 0.5) {
        stop("Class counts do not sum to the sample size")
      }
      fraction_error <- abs(
        sum(vapply(
          year_rows,
          `[[`,
          numeric(1),
          "sample_fraction"
        )) - 1
      )
      maximum_fraction_error <- max(
        maximum_fraction_error,
        fraction_error
      )
      if (fraction_error > 1e-6) {
        stop("Sample fractions do not sum to one")
      }
      if (area_closure_error > glc_sample_closure_tolerance) {
        stop("Sampled class areas do not sum to the watershed area")
      }
    } else if (area_closure_error > glc_exact_boundary_tolerance) {
      stop("Exact class areas exceed the 1% raster-boundary limit")
    }
  }
  list(
    site_id = plan$site$site_id,
    method = plan$method,
    rows = length(rows),
    minimum_sample_n = minimum_sample_n,
    maximum_fraction_sum_error = maximum_fraction_error,
    maximum_area_closure_relative_error = maximum_area_closure_error,
    area_closure_relative_error_limit = if (plan$method == "sample") {
      glc_sample_closure_tolerance
    } else {
      glc_exact_boundary_tolerance
    }
  )
}

### Annual consolidation

annual_glc_settings <- function() {
  list(
    label = "Annual GLC",
    output_stem = "glc_fcs30d_safe_combined",
    normalize = glc_normalize_row,
    validate = glc_validate_asset_rows,
    to_data_frame = function(rows) {
      glc_rows_to_data_frame(rows, glc_output_columns)
    },
    row_key = function(row) {
      paste(row$site_id, row$Year, row$LC_ID, sep = ":")
    },
    expected_rows = function(site_count) {
      site_count * length(glc_years) * length(glc_classes)
    },
    order_by = c("site_id", "Year", "LC_ID")
  )
}

annual_glc_main <- function() {
  args <- glc_parse_args(commandArgs(trailingOnly = TRUE))
  run_glc_consolidation(annual_glc_settings(), args)
}

if (sys.nframe() == 0L) {
  tryCatch(
    annual_glc_main(),
    error = function(error) {
      message("Annual GLC consolidation failed: ", conditionMessage(error))
      quit(status = 1)
    }
  )
}
